// light-charts/lib/utils/animation.js
"use strict";

exports.__esModule = true;
exports.tween = exports.nextFrame = exports.cancelFrame = void 0;
var lasttime = 0;
var nextFrame = window.requestAnimationFrame || window.webkitRequestAnimationFrame || window.mozRequestAnimationFrame || window.msRequestAnimationFrame || function (callback) {
  var curtime = +new Date(),
    delay = Math.max(1000 / 60, 1000 / 60 - (curtime - lasttime));
  lasttime = curtime + delay;
  return setTimeout(callback, delay);
};

// 偷个懒
exports.nextFrame = nextFrame;
var cancelFrame = window.cancelAnimationFrame || window.webkitCancelAnimationFrame || clearTimeout;

// 动画函数
exports.cancelFrame = cancelFrame;
var tween = {
  linear: function linear(t, b, c, d) {
    return c * t / d + b;
  },
  ease: function ease(t, b, c, d) {
    return -c * ((t = t / d - 1) * t * t * t - 1) + b;
  },
  'ease-in': function easeIn(t, b, c, d) {
    return c * (t /= d) * t * t + b;
  },
  'ease-out': function easeOut(t, b, c, d) {
    return c * ((t = t / d - 1) * t * t + 1) + b;
  },
  'ease-in-out': function easeInOut(t, b, c, d) {
    if ((t /= d / 2) < 1) return c / 2 * t * t * t + b;
    return c / 2 * ((t -= 2) * t * t + 2) + b;
  },
  bounce: function bounce(t, b, c, d) {
    if ((t /= d) < 1 / 2.75) {
      return c * (7.5625 * t * t) + b;
    } else if (t < 2 / 2.75) {
      return c * (7.5625 * (t -= 1.5 / 2.75) * t + .75) + b;
    } else if (t < 2.5 / 2.75) {
      return c * (7.5625 * (t -= 2.25 / 2.75) * t + .9375) + b;
    } else {
      return c * (7.5625 * (t -= 2.625 / 2.75) * t + .984375) + b;
    }
  }
};
exports.tween = tween;
